_posts/2019-07-01-test-design.md

```diff
@@ -0,0 +1,294 @@
+
+        return scenario;
+    }
+
+    public CreateOrderScenario modifyCreateOrderData(Consumer<CreateOrderData> modifier) {
+        modifier.accept(createOrderData);
+        return this;
+    }
+
+    public CreateOrderScenario overrideCustomerRepository(Consumer<OrderRepository> modifier) {
+        customerRepository = mock(CustomerRepository.class);
+        modifier.accept(customerRepository);
+        return this;
+    }
+
+    public Order execute() {
+        final OrderValidator orderValidator = new OrderValidator(customerRepository);
+        final OrderFactory orderFactory = new OrderFactory(orderValidator);
+
+        return orderFactory.createOrder(createOrderData);
+    }
+}
+```
+
+In the following example we created two tests that verify that the construction of an order works correctly.
+In the first test, we use the default scenario without modifying anything, meaning, we test the happy path and verify that all data in the created order is correct.
+
+In the second test we verify that if we try to create an order for a customer that doesn't exist, we get a validation exception.
+We do this by creating a default scenario, then modifying the input data to use a customerId defined in the test, and then overriding the behaviour of the CustomerRepository mock.
+
+```java
+class CreateOrderTest {
+    
+    @Test
+    void createOrder_happyPath() {
+        final Order order = CreateOrderScenario
+            .defaultScenario()
+            .execute();
+        
+        assertThat(order.getCustomerId()).isEqualTo(CreateOrderScenario.customerId);
+    }
+    
+    @Test
+    void createOrder_customerInactive() {
+        final UUID customerId = UUID.randomUUID();
+        final CreateOrderScenario scenario = CreateOrderScenario
+            .defaultScenario()
+            .modifyCreateOrderData(orderData -> orderData.setCustomerId(customerId))
+            .overrideCustomerRepository(repository -> 
+                when(repository.findById(customerId)).thenReturn(null));
+        
+        assertThrows(InvalidCustomerException.class, () -> scenario.execute());
+    }
+}
+```
+
+## Advantages of using an API
+
+The advantage of this design is that our tests are not aware of:
+* the use of a factory to create orders,
+* a validator class, used by the factory to validate the input for creating an order,
+* and how the constructor of the Order aggregate should be called. 
+
+With this example, it's easy to see that our new tests are much less likely to break than our original **design**. 
+There is a clean layer between the implementation/design of our application, and the tests.
+This lower coupling makes it easier to refactor the application, and implement new features at a higher pace.
+
+Also notice that we didn't create a test class that maps one-to-one to a production code class.
+Rather than testing our Order object, or our OrderFactory, or our OrderValidator, we test the creation of an aggregate instance. 
```

Comment:
```suggestion
Rather than testing our `Order` object, or our `OrderFactory`, or our `OrderValidator`, we test the creation of an aggregate instance. 
```